cdap-ui/app/cdap/components/PipelineSummary/Store/PipelineSummaryActions.js

```diff
@@ -82,6 +82,81 @@ function updateMetrics(metrics, mapKey = 'metrics') {
   });
 }
 
+function updateNodeMetrics(pipelineConfig) {
+  let {runs} = PipelineSummaryStore.getState().pipelinerunssummary;
+  const sourcePluginTypes = ['batchsource', 'realtimesource', 'streamingsource'];
+  const sinkPluginTypes = ['batchsink', 'realtimesink', 'sparksink'];
+
+  let nodesMap = {
+    sources: {},
+    sinks: {}
+  };
+  pipelineConfig.config.stages.forEach(stage => {
+    if (sourcePluginTypes.indexOf(stage.plugin.type) !== -1) {
+      nodesMap.sources[stage.name] = [];
+    }
+    if (sinkPluginTypes.indexOf(stage.plugin.type) !== -1) {
+      nodesMap.sinks[stage.name] = [];
+    }
+  });
+  /*
+    Sample node metrics:
+      {
+        "user.TPFSAvro.records.out": 40,
+        "user.TPFSAvro.records.in": 40,
+        "user.File.records.in": 40,
+        "user.File.records.out": 40,
+      }
+    Sample nodesMap:
+      {
+        sources: {
+          TPFSAvro: {}
+        },
+        sinks: {
+          File: {}
+        }
+      }
+  */
+  const getRecords = (run, nodesMap, type = '.records.in') => {
+    let cleanUpRegex = /\user.|\.records\.out|\.records\.in/gi;
+    let getNodeLabel = (node) => node.replace(cleanUpRegex, () => '');
+    let nodesCount = Object.keys(nodesMap);
+    let metricsCount = Object.keys(run.nodesMetrics);
+    let defaultMetrics = {};
+    Object.keys(nodesMap).map(n => defaultMetrics[`user.${n}${type}`] = 0);
+    let nodesWithMetrics = run.nodeMetrics;
+    if (!nodesCount.length || nodesCount.length!== metricsCount.length) {
+      nodesWithMetrics = Object.assign({}, defaultMetrics, run.nodesMetrics);
+    }
+    return Object
+      .keys(nodesWithMetrics)
+      .filter(node => node.indexOf(type) !== -1)
+      .filter(node => !isNil(nodesMap[getNodeLabel(node)]))
```

Comment:
can these two filters be combined so that we are not traversing the array twice?